fish-rust/src/curses.rs

```diff
@@ -0,0 +1,300 @@
+        pub fn tgetnum(id: *const libc::c_char) -> libc::c_int;
+
+        pub fn tgetstr(
+            id: *const libc::c_char,
+            area: *mut *mut libc::c_char,
+        ) -> *const libc::c_char;
+    }
+}
+
+pub struct Term {
+    /// A list of cached string values corresponding to known curses string capabilities. Individual
+    /// values may be overridden with [`Term::set()`].
+    ///
+    /// The list is prepopulated on initialization in [`Term::new()`], as such the `Vec` itself is
+    /// never manipulated - only potentially the values within its individual cells.
+    strings: Vec<RefCell<Option<Arc<CString>>>>,
+}
+
+impl Term {
+    /// Looks up support for [`Capability`] `capability` in the termcap/terminfo database via the
+    /// curses library.
+    pub fn get<C: Capability>(&self, capability: C) -> C::Result {
+        capability.lookup(self)
+    }
+
+    /// Overrides the string value of `capability` for the current terminal.
+    pub fn set<S>(&self, id: StringCap, value: Option<S>)
+    where
+        S: ToCString,
+    {
+        let value = value.map(|s| Arc::new(s.to_cstring()));
+        *self.strings[id.idx()].borrow_mut() = value;
+    }
+
+    /// Initialize a new `Term` instance, prepopulating the values of all the curses string
+    /// capabilities we care about in the process.
+    fn new() -> Self {
+        let mut strings = vec![RefCell::new(None); STRING_CAPS.len()];
+        for cap in STRING_CAPS {
+            strings[cap.idx()] = cap
+                .sys_lookup()
+                // Convert the Option<CString> to a RefCell<Option<Arc<CString>>>
+                .map(Arc::new)
+                .into();
+        }
+
+        Term { strings }
+    }
+}
+
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+// Clippy deduces the color of all the cows in Scotland from the first three entries we happen to
+// have here.
+#[allow(clippy::enum_variant_names)]
+enum StringCapIdx {
+    EnterItalicsMode,
+    ExitItalicsMode,
+    EnterDimMode,
+}
+
+// It's preferred **but not required** for the order of `StringCap`s here to match the order of
+// `StringCapIdx` entries, as we always use the associated `StringCapIdx::idx()` to get the index.
+const STRING_CAPS: [StringCap; 3] = [
+    StringCap::new("ZH", StringCapIdx::EnterItalicsMode),
+    StringCap::new("ZR", StringCapIdx::ExitItalicsMode),
+    StringCap::new("mh", StringCapIdx::EnterDimMode),
+];
+
+// String capabilities
+pub const ENTER_ITALICS_MODE: StringCap = STRING_CAPS[StringCapIdx::EnterItalicsMode.idx()];
+pub const EXIT_ITALICS_MODE: StringCap = STRING_CAPS[StringCapIdx::ExitItalicsMode.idx()];
+pub const ENTER_DIM_MODE: StringCap = STRING_CAPS[StringCapIdx::EnterDimMode.idx()];
+
+// Number capabilities
+pub const MAX_COLORS: Number = Number::new("Co");
+
+// Flag capabilities
+pub const EAT_NEWLINE_GLITCH: Flag = Flag::new("xn");
+
+pub trait Capability {
```

Comment:
It's very confusing how some of the Capabilities look at their passed in `Term`, and others look at whatever `cur_term` has. The thread-safety story is completely broken here.

Again I think the simplest and best thing to do is to eagerly load all the capabilities we care about in `setupterm()`, while holding a lock.

Reply:
I didn't realize you wanted to preload numeric and flag capabilities as well. 

I considered the thread story but it's actually "safe" for non-string capabilities, though the results are not defined. If you assume that `cur_term` is swapped out (or zeroed) atomically (and that's a safe assumption on all platforms since the pointer size is guaranteed to be the size of an atomic read) then if that happens the worst that you end up with is that `Term` was initialized for `xterm-256color` but you get back the results for `tmux` (I don't know how curses does it internally and whether there are memory barriers that would prevent stale reads, etc. so maybe there's a chance that you'll end up with one or the other but not both?).